File: odepack/odepack_test.c
#include "odepack.h"
#include <stdlib.h>
#include <stdio.h>
#include <math.h>


void f(int *neq, double *t, double *y, double *ydot) {
    ydot[0] = -y[0];
    ydot[1] = -y[1];
}

void g(int *neq, double *t, double *y, int *ng, double *gout) {
    gout[0] = *t - 3.;
}

#define SIZE(x) (sizeof(x)/sizeof(x[0]))

int main() {

    int i;
    double t[] = {0., 1., 2., 3., 4., 5., 6.};
    double y[SIZE(t)*2] = {1., 1.};

    data DATA = {0};
    data *D = &DATA;

    D->f = f;
    D->neq = 2;
    D->n = SIZE(t);
    D->t = t;
    D->y = y;

    D->ng = 1;
    D->g = g;

    odesolve(D);

    for (i=0; i<D->points_done; i++) {
        printf("%e, %e, %e\n", y[2*i], y[2*i+1], exp(-t[i]));
    }

    return 0;
}

File: odepack/python/odepack_test.py
import numpy as np
from odepack import odesolve

neq = 2
n = 6
ng = 1

def f(y, t, ydot):
    ydot[0] = -y[0]
    ydot[1] = -y[1]

y = np.ones((n, neq))
t = np.arange(n, dtype=np.float64)

work = np.zeros(neq), np.zeros(1), np.zeros(neq), np.zeros(ng)

points_done = odesolve(f, y, t, work)
t1, y1 = t[:points_done], y[:points_done]

print y1[:,0]
print np.exp(-t1)


def g(y, t, gout):
    gout[0] = y[0] - 0.2

points_done = odesolve(f, y, t, work, g)
t2, y2 = t[:points_done], y[:points_done]

print y2[:,0]
print np.exp(-t2)


ibbox = np.array([0], 'i')
bbox = np.array([0.2])

points_done = odesolve(f, y, t, work, None, ibbox, bbox)
t3, y3 = t[:points_done], y[:points_done]

print y3[:,0]
print np.exp(-t3)


